testing/Android_SDK_E2E_Test suite.md

```diff
@@ -0,0 +1,71 @@
+
+### Webdriverio (nodejs)
+
+Like the WD.js library like the WD.js client, tests using this client can be written in a short chained syntax and can also be used in conjunction with features of tools like mochas transfer promisness to make assertions part of this chained syntax.  Unlike the WD.js client however, webdriverio can handle the redirect to the browser from the application, however there is only two possible locator strategies - xpath and accessibilityId. 
+
+### Java-Client (Appium Java SDK)
+
+Unlike the javascript libraries the java-client cannot be written in a promise chain syntax, however, there is a large number of locator strategies and it can handle the transition between native application and mobile browser.  Another major feature of the java client is the support for page object classes - using the following annotations, @FindBy(someStrategy), @AndroidFindBy(someStrategy) and @iOSFindBy(someStrategy), allows the creation of a single page object when creating tests for applications on multiple platforms that have the same content and flow but identifiers unique to their platform rather than a separate page object for each [3]. 
+
+## Framework Recommendation
+
+Appium is the framework of choice from those explored as it is the only framework explored that provides clients that can handle the redirect to the browser that is an integral test case we must cover.  Of the two clients, the Java-Client is likely to be the superior choice as it has a greater number of locator strategies and the annotation selector targeting provided can reduce the size of the codebase.
+
+## Test suite Proposal
+
+### Implementation Language
+
+With the selection of the Appium framework and its Java-Client java, or a JVM based language such as Kotlin, Groovy, Scala, etc. should be used, provided they can support the java based annotations that the appium Java-Client provides for Page Object support.  The use of a JVM language other than Java could increase the readability of the test suite.
+
+### Page Objects
+
+The test suite should make use of the page object model. This model separates the logic the presentation logic of the associated page objects and the test logic which leads to tests that are less likely to be brittle, and would therefore increase the maintainability of the test suite [4].  As mentioned in the framework recommendation section, the ability of the Appium library to use annotations for targeting selectors that are unique to each platform can help to reduce the codebase size - only one page object is required per platform with the annotations providing the ability to target Android or iOS specific selectors and the tests files themselves can also be generic meaning no separate tests for Android or iOS applications that share the exact same pattern of logic flow through the application.
```

Comment:
IMHO - that's too low level for proposal. It's just implementation specifics that may depend on solutions we going to select. 
Initially we just need to know - what is going to be tested, what tooling will be created etc.

Reply:
@wtrocki 

> I think that main risk for this is that at some point it may happen that examples may diverge because of platform specifics and we will not be able to use single test suite. Making sure that both example apps will work with the same test suite will require extra effort etc.

We can still create tests specific to each as well, but for anything that is common we could have common tests. There is a possibility to write our tests to cover multiple os platforms which we are highlighting here.

There should also be no extra effort to make the apps work from the app development side, but I am assuming there is some sort of planned common UI flow logic and layout. The specifics of things like ids, classes, etc. do not matter.

> Another smaller risk is IOS, Cordova, React, JS and Xamarin support.

There is no problem in support. Appium can test native, hybrid and mobile web apps on iOS, Android, and Windows mobile devices. A quick web search will show tutorials of how these are achieved in the technologies you have suggested.

> IMHO - that may too low level for proposal.
Initially we just need to know - what is going to be tested, what tooling will be created etc.
It will be good to present just top level overview.

In my view a high level is fine if you are starting from scratch, but there has been a lot of investigation into frameworks done already. With this information the proposal should then outline the pros and cons of each framework (done), and produce a reccomendation (done).

While we started out on considering the Android first, Appium can test any mobile apps built from the SDKs we plan to create and more. Buit Android is our starting point.

When writing the proposal for e2e testing this recommendation is how I have interpreted the framework/test suite to be created from investigations we undertook, discussions with @wojta and adhering to best practices like page objects. Specific details of how page objects can be created are there to illustrate the possibilities in reducing codebase size and increasing maintainability but may end up not being used in the actual implemented code.